src/Microsoft.Graph.Core/Requests/RedirectHandler.cs

```diff
@@ -0,0 +1,159 @@
+                    }
+                    else
+                    {
+                        newRequest.Content = content;
+                        newRequest.Method = request.Method;
+                    }
+
+                    // Set newRequestUri from response
+                    newRequest.RequestUri = response.Headers.Location;
+
+                    // Remove Auth if unneccessary
+                    if (String.Compare(newRequest.RequestUri.Host, request.RequestUri.Host, StringComparison.OrdinalIgnoreCase) != 0)
+                    {
+                        newRequest.Headers.Authorization = null;
+
+                    }
+
+                    // Send redirect request to get reponse      
+                    response = await base.SendAsync(newRequest, cancellationToken);
+
+                    // Check response status code
+                    if (!IsRedirect(response.StatusCode))
+                    {
+                        return response;
+                    }
+                    redirectCount++;
+                }
+                throw new ServiceException(
+                        new Error
+                        {
+                            Code = ErrorConstants.Codes.TooManyRedirects,
+                            Message = string.Format(ErrorConstants.Messages.TooManyRedirectsFormatString, redirectCount)
+                        });
+
+            }
+            return response;
+
+        }
+
+        /// <summary>
+        /// Copy original HTTP request's headers and porperties.
+        /// </summary>
+        /// <param name="originalRequest">The original <see cref="HttpRequestMessage"/> needs to be copy.</param>
+        /// <returns>The <see cref="HttpRequestMessage"/>.</returns>
+        /// <remarks>
+        /// Re-issue a new HTTP request with the original request's headers and properities
+        /// </remarks>
+        internal HttpRequestMessage CopyRequest(HttpRequestMessage originalRequest)
+        {
+            var newRequest = new HttpRequestMessage(originalRequest.Method, originalRequest.RequestUri);
+
+            foreach (var header in originalRequest.Headers)
+            {
+                newRequest.Headers.TryAddWithoutValidation(header.Key, header.Value);
+            }
+
+            foreach (var property in originalRequest.Properties)
+            {
+                newRequest.Properties.Add(property);
+            }
+
+            return newRequest;
+        }
+
+
+        /// <summary>
+        /// Checks whether <see cref="HttpStatusCode"/> needs redirected
+        /// </summary>
+        /// <param name="statusCode">The <see cref="HttpStatusCode"/>.</param>
+        /// <returns>Bool value for redirection or not</returns>
+        private bool IsRedirect(HttpStatusCode statusCode)
+        {
+            return (int)statusCode >= 300 && (int)statusCode < 400 && statusCode != HttpStatusCode.NotModified && statusCode != HttpStatusCode.UseProxy;
```

Comment:
I would rather us be explicit here about the status codes.  Codes like 304 and 305 are not actually redirects.

Reply:
Thanks. will update it.

Reply:
Yes, it actually works with not redirect codes like 304 and 305. But be explicit is better. I updated the method with explicit status codes. Btw, since System.Net.Http.HttpStatusCode Enum doesn't contain 308 in .NET Framework 4.7.2 version now, I used explicit status code 308 in the method. I'm wondering would the code below
var response = await base.SendAsync(request, cancellationToken); 
throws an exception if the client returned with 308 response?

Reply:
@vikyding No exception will be thrown from SendAsync for any HTTP response.  Only broken connections and timeouts would cause exceptions to be thrown.